In [1]:
import numpy as np
from scipy.stats import t

## 📌 Question 1 : Paired t-Test (Blood Pressure)

Blood pressure (mmHg) measured before and after medication for 8 patients:  
Before: `[140, 138, 150, 145, 142, 148, 135, 152]`  
After: `[132, 130, 142, 138, 135, 140, 128, 144]`  
At α = 0.05, test whether the medication significantly reduces blood pressure. Also compute a 95% CI for the mean reduction.

#### Step 1 : Hypotheses

Let $\mu_d$ = true mean of (before - after) differences.

$$H_0: \mu_d = 0 \quad \text{(no change)}$$
$$H_1: \mu_d > 0 \quad \text{(blood pressure decreased, one-tailed)}$$

#### Step 2 : Why this test

The same 8 patients are measured twice, so the observations are paired. A paired t-test reduces between-subject noise by working directly on the within-subject differences. With the differences treated as a single sample, this collapses to a one-sample t-test on $d_i = \text{before}_i - \text{after}_i$. Population variance is unknown, so t (not Z) is correct.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{s_d / \sqrt{n}}, \quad df = n - 1$$

where $\bar{d}$ is the mean difference, $s_d$ is the sample standard deviation of the differences.

In [2]:
before = np.array([140, 138, 150, 145, 142, 148, 135, 152])  # before readings
after  = np.array([132, 130, 142, 138, 135, 140, 128, 144])  # after readings
alpha  = 0.05                                                  # significance level
n      = len(before)                                           # sample size

In [3]:
d     = before - after          # paired differences
d_bar = d.mean()                # mean difference
sd    = np.std(d, ddof=1)       # sample std of differences
se    = sd / np.sqrt(n)         # standard error
df    = n - 1                   # degrees of freedom
t_stat = d_bar / se             # test statistic

#### Step 4 : Critical Value Method

In [4]:
t_crit = t.ppf(1 - alpha, df)          # one-tailed critical value
print(f't_stat : {t_stat:.3f}')
print(f't_crit : {t_crit:.3f}')

t_stat : 41.671
t_crit : 1.895


**t_stat (41.67) > t_crit (1.895) → Reject H₀**

#### Step 5 : p-value Method

In [5]:
pval = t.sf(t_stat, df)                # one-tailed p-value (upper tail)
print(f'p-val : {pval:.12f}')

p-val : 0.000000000598


**p-val ≈ 0.0000000006 < 0.05 → Reject H₀**

#### Step 6 : Confidence Interval

A 95% CI for $\mu_d$ uses the two-tailed critical value:

$$\bar{d} \pm t_{\alpha/2,\, df} \cdot \frac{s_d}{\sqrt{n}}$$

In [6]:
t_ci  = t.ppf(1 - alpha/2, df)         # two-tailed critical value for CI
moe   = t_ci * se                       # margin of error
lo    = d_bar - moe                     # lower bound
hi    = d_bar + moe                     # upper bound
print(f'95% CI : ({lo:.3f}, {hi:.3f})')

95% CI : (7.192, 8.058)


#### ✅ Final Conclusion

| Metric | Value |
|---|---|
| Mean difference $\bar{d}$ | 7.625 mmHg |
| Std deviation $s_d$ | 0.518 |
| Standard error | 0.183 |
| t statistic | 41.671 |
| t critical (one-tailed) | 1.895 |
| p-value | ≈ 0.0000000006 |
| 95% CI for $\mu_d$ | (7.192, 8.058) |
| Decision | Reject H₀ |

Strong evidence that the medication reduces blood pressure. The mean reduction is about 7.6 mmHg, and the 95% CI (7.19, 8.06) sits entirely above zero, confirming the effect is real and practically meaningful.

---

## 📌 Question 2 : Paired t-Test + Effect Size (VO₂ Max)

VO₂ max differences after a 12-week fitness program for 10 athletes:  
`[+4.2, +3.8, +5.1, +2.9, +6.0, +3.3, +4.7, +5.5, +2.1, +4.4]`  
Test at α = 0.05 (one-tailed) whether the program improved VO₂ max.  
Also compute Cohen's $d_z = \bar{d} / s_d$ and interpret the effect size.

#### Step 1 : Hypotheses

Let $\mu_d$ = true mean improvement in VO₂ max.

$$H_0: \mu_d = 0 \quad \text{(no improvement)}$$
$$H_1: \mu_d > 0 \quad \text{(program improved VO₂ max, one-tailed)}$$

#### Step 2 : Why this test

The differences are already provided directly (before minus after is implicit). Each value is the within-subject change for one athlete, so this is still a paired design. With $n = 10$ and unknown population variance, a one-sample t-test on the differences is appropriate. One-tailed because the question only asks about improvement.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{s_d / \sqrt{n}}, \quad df = n - 1$$

In [7]:
d      = np.array([4.2, 3.8, 5.1, 2.9, 6.0, 3.3, 4.7, 5.5, 2.1, 4.4])  # VO2 max improvements
n      = len(d)                                                             # sample size
d_bar  = d.mean()                                                           # mean improvement
sd     = np.std(d, ddof=1)                                                  # sample std
se     = sd / np.sqrt(n)                                                    # standard error
df     = n - 1                                                              # degrees of freedom
t_stat = d_bar / se                                                         # test statistic

#### Step 4 : Critical Value Method

In [8]:
t_crit = t.ppf(1 - alpha, df)          # one-tailed critical value
print(f't_stat : {t_stat:.3f}')
print(f't_crit : {t_crit:.3f}')

t_stat : 11.009
t_crit : 1.833


**t_stat (11.009) > t_crit (1.833) → Reject H₀**

#### Step 5 : p-value Method

In [9]:
pval = t.sf(t_stat, df)                # one-tailed p-value (upper tail)
print(f'p-val : {pval:.12f}')

p-val : 0.000000799635


**p-val ≈ 0.0000008 < 0.05 → Reject H₀**

#### Step 6 : Confidence Interval

95% CI for $\mu_d$:

$$\bar{d} \pm t_{\alpha/2,\, df} \cdot \frac{s_d}{\sqrt{n}}$$

In [10]:
t_ci  = t.ppf(1 - alpha/2, df)         # two-tailed critical value for CI
moe   = t_ci * se                       # margin of error
lo    = d_bar - moe                     # lower bound
hi    = d_bar + moe                     # upper bound
print(f'95% CI : ({lo:.3f}, {hi:.3f})')

95% CI : (3.337, 5.063)


#### Part (b) : Cohen's $d_z$ (Effect Size)

$$d_z = \frac{\bar{d}}{s_d}$$

Benchmarks: 0.2 = small, 0.5 = medium, 0.8 = large.

In [11]:
d_z = d_bar / sd                        # Cohen's d_z for paired design
print(f"Cohen's d_z : {d_z:.3f}")

Cohen's d_z : 3.481


#### ✅ Final Conclusion

| Metric | Value |
|---|---|
| Mean improvement $\bar{d}$ | 4.200 mL/kg/min |
| Std deviation $s_d$ | 1.206 |
| Standard error | 0.382 |
| t statistic | 11.009 |
| t critical (one-tailed) | 1.833 |
| p-value | ≈ 0.0000008 |
| 95% CI for $\mu_d$ | (3.337, 5.063) |
| Cohen's $d_z$ | 3.481 |
| Decision | Reject H₀ |

The program significantly improved VO₂ max. The mean gain of 4.2 mL/kg/min is both statistically significant and practically large. Cohen's $d_z \approx 3.48$ is far above the "large" threshold of 0.8, meaning the treatment effect is enormous relative to within-subject variability.